doc/ACL_pytest.md

```diff
@@ -0,0 +1,326 @@
+def config(request, dut, setup):
+    '''
+    Fixture generates ACL tables, ACL rules configuration and deploys on the DUT
+    The return value is config files generated
+    '''
+
+@pytest.yield_fixture(scope='module')
+def table(dut, config, stage):
+    '''
+    Create an ACL table on DUT
+    '''
+
+```
+
+Based on the above it is assumed that tests in one module will reuse the ACL table created by 'table' fixture.
+
+If, for example, some one is developing ACL table scale test he will need a seperate test module ('test_acl_scale.py') and override 'config' fixutre to generate many tables configuration and reuse 'table' fixture.
+
+###### Class scope
+
+
+```python
+
+class BaseAclTest(object):
+    __meta__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def setup_rules(self):
+        pass
+
+    @abc.abstractmethod
+    def teardown_rules(self):
+        pass
+
+    @pytest.yield_fixture(scope='class')
+    def rules(self, dut, table):
+        '''
+        Setup/Teardown ACL rules on DUT
+        '''
+
+    @pytest.yield_fixture(scope='class', autouse=True)
+    def counters_sanity_check(self, dut, rules):
+        '''
+        Sanity checks for ACL counters: verify ACL counters increased
+        '''
+
+    def test_traffic(self):
+        # ...
+
+```
+
+
+#### Isolate test cases in PTF script *acltb_test.py*
+
+Current implementation of the PTF test includes all the test cases under signle test object. There are two issue with current approach
+- Test cases are not isolated, so user cannot run individual test without modifying sources mannualy
+- Test cases hardcode ACL rule behaviour, this means we have ACL rule configuration generation step which defines which rules to test and PTF test script must be aligned with configuration; we will try to have PTF test very generic for different ACL rules and py.test logic will invoke PTF with parameters aligned with generated ACL configuration
+
+PTF test script will have a base class with defined methods to send/receive:
+
+```python
+class AclRuleTest(BaseTest):
+    def setUp(self):
+        # generic parameters from self.test_params
+        pass
+
+    def runTest(self):
+        # run test
+
+class L4SourcePortMatch(AclRuleTest):
+    def setUp(self):
+        super(L4PortRangeMatch, self).setUp()
+        self.sport = self.test_params['source_port']
+        self.pkt.sport = self.exp_pkt.sport = self.sport
+```
+
+Invokation like this:
+
+```
+ptf \
+    --test-dir acstests acltb_test.L4SourcePortMatch
+    --platform-dir ptftests
+    --platform remote
+    -t "
+        # ...
+        proto='udp',
+        source_port='4126';
+        direction='spine->tor';
+        forward='True';
+       "
+```
+
+Will run the test for L4 source port match ACL rule wich is expected to accept traffic
+
+###### Pytest code
+
+In py.test script the logic will look like this:
+
+```python
+
+class BaseAclTest(object):
+
+    @pytest.mark.parametrize("test_case, proto, forward, direction, options",
```

Comment:
There are cases for testing ACL priority. Multiple ACL rules with different priority need to be loaded. Is this pattern able to meet that requirement?

Reply:
All ACL rules are created in scope of a class running the traffic cases. Yes this will meat the requirement.

Lets say '# Test #20 - rules priority - forwarded'

```python
('SourceIpMatch', 'TCP', 'forward', 'spine->tor', { 'source_ip' = '20.0.0.7' } )
```

One thing I haven't consider - in the test run output it won't be clear that we test rules priority.
Maybe I will add 'short_description_field' so that in test output the test case intention is more clear.

In general I am thinking which cases of combinations of matches and actions in future this pattern can cover. Right now only actions are drop and forward, but there might be more I need to consider